File: question4/src/main/java/com/example/quesion4/myAdapter.java
package com.example.quesion4;

import android.content.Context;
import android.graphics.Color;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class myAdapter extends RecyclerView.Adapter<myAdapter.MyViewHolder> {
    private static final String TAG = "myAdapter";
    private OnClickListenerCallback onClickListenerCallback;

    myAdapter(OnClickListenerCallback onClickListenerCallback) {
        this.onClickListenerCallback = onClickListenerCallback;
    }


    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View myView = LayoutInflater.from(parent.getContext()).inflate(R.layout.mylayout, parent, false);
        return new MyViewHolder(myView);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, final int position) {
        holder.bindData(position, onClickListenerCallback);

    }

    @Override
    public int getItemCount() {
        return onClickListenerCallback.getList().size();
    }


    /**
     * ViewHolder inner-class
     */
    static class MyViewHolder extends RecyclerView.ViewHolder {
        TextView textview;
        ConstraintLayout layout;

        MyViewHolder(@NonNull View itemView) {
            super(itemView);
            layout = itemView.findViewById(R.id.linearLayout);
            textview = itemView.findViewById(R.id.name);
        }

        void bindData(final int position, final OnClickListenerCallback onClickListenerCallback) {
            textview.setText(onClickListenerCallback.getList().get(position).getName());
            layout.setBackgroundColor(Color.parseColor(onClickListenerCallback.getList().
                    get(position).getColorCode()));
            layout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onClickListenerCallback.onClickCallBack(position);
                }
            });
        }
    }
}

File: exercise_14_android_introduction_to_jetpack_component_part_2/src/main/java/com/example/exercise_14_android_introduction_to_jetpack_component_part_2/second_question/adapter/MyAdapter.kt
package com.example.exercise_14_android_introduction_to_jetpack_component_part_2.second_question.adapter

import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.exercise_14_android_introduction_to_jetpack_component_part_2.R
import com.example.exercise_14_android_introduction_to_jetpack_component_part_2.second_question.interfaces.OnClickListenerColorChange
import com.example.exercise_14_android_introduction_to_jetpack_component_part_2.second_question.model.ColorList
import kotlinx.android.synthetic.main.color_layout.view.*
import java.util.*

/**
 * RecyclerView Adapter
 */
class MyAdapter internal constructor(private val listener: OnClickListenerColorChange) :
    RecyclerView.Adapter<MyAdapter.MyViewHolder>() {
    private var colors: List<ColorList?> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val myView: View =
            LayoutInflater.from(parent.context).inflate(R.layout.color_layout, parent, false)
        return MyViewHolder(
            myView
        )
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        holder.dataBind(colors[position], listener)
    }

    override fun getItemCount(): Int {
        return colors.size
    }

    fun setList(colors: List<ColorList?>) {
        this.colors = colors
    }

    /**
     * ViewHolder Class
     */
    class MyViewHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        fun dataBind(color: ColorList?, listener: OnClickListenerColorChange) {
            itemView.color_text.text = color!!.name.toUpperCase()
            itemView.color_layout.setBackgroundColor(Color.parseColor(color.code))
            itemView.color_layout.setOnClickListener { listener.onItemClick(color) }
        }
    }
}